File: backend/services/predictor_datasets.py
"""
predictor_datasets.py — Loads all 9 predictor CSVs using pandas.
Mirrors services/predictorDatasets.js.
"""
import os
from pathlib import Path
from typing import Optional

import pandas as pd

DATA_DIR = Path(__file__).parent.parent / "data"

DATASET_MAP = {
    "predict_heart.csv":        "diabetes-heart",
    "predict_dengue.csv":       "dengue",
    "predict_kidney.csv":       "kidney",
    "predict_liver.csv":        "liver",
    "predict_lung.csv":         "lung",
    "predict_cancer.csv":       "cancer",
    "predict_thyroid.csv":      "thyroid",
    "predict_asthma.csv":       "asthma",
    "predict_mental_health.csv": "mental-health",
    "diabetes.csv":             "diabetes-heart",  # extra source
}

_stats: dict = {}
_loaded: bool = False


def load_all() -> dict:
    global _stats, _loaded
    if _loaded:
        return _stats

    total_rows = 0
    for filename, predictor_id in DATASET_MAP.items():
        filepath = DATA_DIR / filename
        if not filepath.exists():
            continue
        try:
            df = pd.read_csv(filepath, low_memory=False)
            rows = len(df)
            columns = list(df.columns)
            if predictor_id not in _stats:
                _stats[predictor_id] = {"rows": 0, "columns": [], "files": []}
            _stats[predictor_id]["rows"] += rows
            existing_cols = set(_stats[predictor_id]["columns"])
            for c in columns:
                if c not in existing_cols:
                    _stats[predictor_id]["columns"].append(c)
                    existing_cols.add(c)
            _stats[predictor_id]["files"].append(filename)
            total_rows += rows
        except Exception as e:
            print(f"[PredictorDatasets] Error loading {filename}: {e}")

    _loaded = True
    print(f"[PredictorDatasets] ✅ Loaded {len(_stats)} predictors — {total_rows:,} total patient records")
    return _stats


def get_stats() -> dict:
    return _stats

File: backend/services/predictorDatasets.js
/**
 * predictorDatasets.js
 * Loads all 9 predictor CSVs into memory once on startup.
 * Provides real statistics (patient count, column info) to enrich AI prompts.
 */

const fs       = require('fs');
const path     = require('path');
const readline = require('readline');

const DATA_DIR = path.join(__dirname, '../data');

// Dataset configs: file → predictor id
const DATASET_MAP = {
  'predict_heart.csv':        'diabetes-heart',
  'predict_dengue.csv':       'dengue',
  'predict_kidney.csv':       'kidney',
  'predict_liver.csv':        'liver',
  'predict_lung.csv':         'lung',
  'predict_cancer.csv':       'cancer',
  'predict_thyroid.csv':      'thyroid',
  'predict_asthma.csv':       'asthma',
  'predict_mental_health.csv':'mental-health',
};

// Also include existing diabetes.csv for diabetes-heart
const EXTRA = {
  'diabetes.csv': 'diabetes-heart',
};

const stats = {}; // { predictorId: { rows, columns, sample } }
let loaded  = false;

function countCSVRows(filePath) {
  return new Promise((resolve) => {
    let count = 0;
    let header = null;
    const rl = readline.createInterface({
      input: fs.createReadStream(filePath, { encoding: 'utf8' }),
      crlfDelay: Infinity,
    });
    rl.on('line', (line) => {
      if (!header) { header = line; return; }
      if (line.trim()) count++;
    });
    rl.on('close', () => resolve({ rows: count, header }));
    rl.on('error', () => resolve({ rows: 0, header: '' }));
  });
}

async function loadAll() {
  if (loaded) return stats;

  const allFiles = { ...DATASET_MAP, ...EXTRA };

  const promises = Object.entries(allFiles).map(async ([filename, predictorId]) => {
    const filePath = path.join(DATA_DIR, filename);
    if (!fs.existsSync(filePath)) return;

    const { rows, header } = await countCSVRows(filePath);
    const columns = header ? header.split(',').map(c => c.trim()) : [];

    // Merge with existing (diabetes-heart has two files)
    if (!stats[predictorId]) {
      stats[predictorId] = { rows: 0, columns: [], files: [] };
    }
    stats[predictorId].rows    += rows;
    stats[predictorId].columns  = [...new Set([...stats[predictorId].columns, ...columns])];
    stats[predictorId].files.push(filename);
  });

  await Promise.all(promises);
  loaded = true;

  const total = Object.values(stats).reduce((s, v) => s + v.rows, 0);
  console.log(`[PredictorDatasets] ✅ Loaded stats for ${Object.keys(stats).length} predictors — ${total.toLocaleString()} total patient records`);

  return stats;
}

// Start loading in background
loadAll().catch(err => console.error('[PredictorDatasets] Load error:', err.message));

module.exports = { loadAll, getStats: () => stats };
